README.md
# docker-machine-driver-outscale

[![Go Report Card](https://goreportcard.com/badge/github.com/outscale-dev/docker-machine-driver-outscale)](https://goreportcard.com/report/github.com/outscale-dev/docker-machine-driver-outscale)
[![GitHub release](https://img.shields.io/github/release/outscale-dev/docker-machine-driver-outscale.svg)](https://github.com/outscale-dev/docker-machine-driver-outscale/releases/)
[![Project Graduated](https://docs.outscale.com/fr/userguide/_images/Project-Graduated-green.svg)](https://docs.outscale.com/en/userguide/Open-Source-Projects.html)

Outscale Driver plugin for docker-machine

## Install
If you would rather build from source, you will need to have a working `go` 1.17+ environment,

```bash
eval $(go env)
export PATH="$PATH:$GOPATH/bin"
```

You can then install `docker-machine` from source by running:

```bash
go get github.com/docker/machine
cd $GOPATH/src/github.com/docker/machine
make build
```

And then compile the `docker-machine-driver-outscale` driver:

```bash
go get github.com/outscale-dev/docker-machine-driver-outscale
cd $GOPATH/src/github.com/outscale-dev/docker-machine-driver-outscale
make install
```

## Run
In order to create a machine, you will need to have you AK/SK . You can find information here: [Documentation](https://docs.outscale.com/en/userguide/Getting-Information-About-Your-Access-Keys.html)

```bash
docker-machine create -d outscale --outscale-access-key=<outscale-access-key>  --outscale-secret-key=<outscale-secret-key> --outscale-region=<outscale-region> outscale
```

### Options
| Argument | Env | Default | Description
| --- | --- | --- | ---
| `outscale-access-key` | `OUTSCALE_ACCESS_KEY\|OSC_ACCESS_KEY` | None | **required** Outscale Access Key (see [here](https://docs.outscale.com/en/userguide/Getting-Information-About-Your-Access-Keys.html))
| `outscale-secret-key` | `OUTSCALE_SECRET_KEY\|OSC_SECRET_KEY` | None | **required** Outscale Secret Key (see [here](https://docs.outscale.com/en/userguide/Getting-Information-About-Your-Access-Keys.html))
| `outscale-region` | `OUTSCALE_REGION\|OSC_REGION` | eu-west-2 | Outscale Region
| `outscale-instance-type` | `OUTSCALE_INSTANCE_TYPE` | tinav2.c1r2p3 (t2.small) | Outscale VM Instance Type (see [here](https://docs.outscale.com/en/userguide/Instance-Types.html))
| `outscale-source-omi`    | `OUTSCALE_SOURCE_OMI`    | ami-2cf1fa3e (Debian-10-2021.05.12-3) | Outscale Machine Image to use as bootstrap for the VM (see [here](https://docs.outscale.com/en/userguide/Official-OMIs-Reference.html#_supported_official_images)) |
| `outscale-extra-tags-all` | `` | nil| Extra tags for all created resources. Format "key=value". Can be set multiple times
| `outscale-extra-tags-instances` | `` | nil | Extra tags only for instances. Format "key=value". Can be set multiple times
| `outscale-security-group-ids` | `` | nil | Ids of user defined Security Groups to add to the machine. Can be set multiple times
| `outscale-root-disk-type` | `` | gp2 | Type of volume for the root disk ('standard', 'io1' or 'gp2')
| `outscale-root-disk-size` | `` | 15 | Size of the root disk in GB (> 0)
| `outscale-root-disk-iops` | `` | 1500 | Iops for the io1 root disk type (ignore if it is not io1). Value between 1 and 13000.
| `outscale-subnet-id` | `` | `` | Id of the Net use to create all resources when a private network is requested.
| `outscale-kubernetes-node-name-autotag` | `` | false | Automatically add kubernetes tag 'OscK8sNodeName' to the instance (Useful for the CCM).


## Security group
If no Security group is provided, a security group will be created with theses rules
| Type | Protocol | From Port | To Port | CIDR | Description
| --- | --- | --- | --- | --- | ---
| Inbound | TCP | 22 | 22 | 0.0.0.0/0 | SSH
| Inbound | TCP | 80 | 80 | 0.0.0.0/0 | nginx Ingress Http
| Inbound | TCP | 443 | 443 | 0.0.0.0/0 | nginx Ingress Https
| Inbound | TCP | 2376 | 2376 | 0.0.0.0/0 | Docker daemon
| Inbound | TCP | 2379 | 2380 | 0.0.0.0/0 | ETCD (client request and peer communication)
| Inbound | TCP | 6443 | 6443 | 0.0.0.0/0 | Kube-api 
| Inbound | TCP | 10250 | 10250 | 0.0.0.0/0 | Kubelet
| Inbound | TCP | 10251 | 10251 | 0.0.0.0/0 | Kube-scheduler
| Inbound | TCP | 10252 | 10252 | 0.0.0.0/0 | Kube-controller-manager
| Inbound | TCP | 10256 | 10256 | 0.0.0.0/0 | Kube-proxy
| Inbound | TCP | 30000 | 32767 | 0.0.0.0/0 | Node port
| Inbound | UDP | 30000 | 32767 | 0.0.0.0/0 | Node port
| Inbound | UDP | 8472 | 8472 | 0.0.0.0/0 | Canal/Flannel overlay
| Inbound | UDP | 4789 | 4789 | 0.0.0.0/0 | Canal/Flannel overlay

In the example section, there are some exampe of minimal Security Group preprovisionned for different use-cased:
- [Rancher Cluster with calico network](example/calico/README.md)
- [Rancher Cluster with canal network](example/canal/README.md)

## Debugging
Detailed run output will be emitted when using  the `docker-machine` `--debug` option.

```bash
docker-machine --debug  create -d outscale --outscale-access-key=<outscale-access-key>  --outscale-secret-key=<outscale-secret-key> --outscale-region=<outscale-region> outscale
```

## License

> Copyright Outscale SAS
>
> BSD-3-Clause

This project is compliant with [REUSE](https://reuse.software/).


docs/CHANGELOG.md
# Changelog

## [v0.2.0]
### Features
* Improve error messages ([#27](https://github.com/outscale-dev/docker-machine-driver-outscale/pull/27))
* Support Private Cloud ([#29](https://github.com/outscale-dev/docker-machine-driver-outscale/pull/29))
* Adapt Throttling to Outscale API ([#30](https://github.com/outscale-dev/docker-machine-driver-outscale/pull/30))
* Support autotag of instance for Kubernetes Outscale CCM ([#32](https://github.com/outscale-dev/docker-machine-driver-outscale/pull/32))
### Bugfixes
* Fix SK initialisation ([#31](https://github.com/outscale-dev/docker-machine-driver-outscale/pull/31))


## [v0.1.1]
### Features
* Replace default OMI
* Change root disk iops option name for Rancher

## [v0.1]
### Features
* Add Root disk options

## [v0.0.2]
### Features
* Rename all environment variables to OUTSCALE
* Retry when throttling occur
* Support tag on resources and vms
* Support user-defined security groups

### Bugfixes
* Ensure that no resources leak occurs

## [v0.0.1]
### Features
* Create machine in public cloud


pkg/drivers/outscale/outscale.go
package outscale

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"os"

	retry "github.com/avast/retry-go"
	"github.com/docker/machine/libmachine/drivers"
	"github.com/docker/machine/libmachine/log"
	"github.com/docker/machine/libmachine/mcnflag"
	"github.com/docker/machine/libmachine/state"
	osc "github.com/outscale/osc-sdk-go/v2"
)

const (
	defaultOscRegion       = "eu-west-2"
	defaultOscOMI          = "ami-2cf1fa3e"  // Debian 10
	defaultOscVmType       = "tinav2.c1r2p3" //t2.small
	defaultDockerPort      = 2376
	defaultSSHPort         = 22
	defaultSSHUsername     = "outscale"
	defaultRootDiskType    = "gp2"
	defaultRootDiskSize    = 15
	defaultRootDiskIo1Iops = 1500

	flagAccessKey          = "outscale-access-key"
	flagSecretKey          = "outscale-secret-key"
	flagRegion             = "outscale-region"
	flagInstanceType       = "outscale-instance-type"
	flagSourceOmi          = "outscale-source-omi"
	flagExtraTagsAll       = "outscale-extra-tags-all"
	flagExtraTagsInstances = "outscale-extra-tags-instances"
	flagSecurityGroupIds   = "outscale-security-group-ids"
	flagRootDiskType       = "outscale-root-disk-type"
	flagRootDiskSize       = "outscale-root-disk-size"
	flagRootDiskIo1Iops    = "outscale-root-disk-iops"
	flagSubnetId           = "outscale-subnet-id"
	flagK8sNodeNameTag     = "outscale-kubernetes-node-name-autotag"
)

type OscDriver struct {
	*drivers.BaseDriver

	oscApi *OscApiData

	// Stored
	Ak     string
	Sk     string
	Region string

	VmId            string
	KeypairName     string
	SecurityGroupId string
	PublicIpId      string
	PublicCloud     bool

	// Unstored
	instanceType       string
	sourceOmi          string
	extraTagsAll       []string
	extraTagsInstances []string
	securityGroupIds   []string
	rootDiskType       string
	rootDiskSize       int32
	rootDiskIo1Iops    int32
	subnetId           string
	netId              string
	tagK8sNodeName     bool
}

type OscApiData struct {
	client  *osc.APIClient
	context context.Context
}

// NewDriver creates and returns a new instance of the Outscale driver
func NewDriver(hostName, storePath string) *OscDriver {
	return &OscDriver{
		BaseDriver: &drivers.BaseDriver{
			MachineName: hostName,
			StorePath:   storePath,
		},
	}
}

func (d *OscDriver) getClient() (*OscApiData, error) {
	if d.oscApi == nil {
		config := osc.NewConfiguration()

		config.Debug = true
		config.UserAgent = fmt.Sprintf("docker-machine-driver-outscale/%s", GetVersion())

		client := osc.NewAPIClient(config)

		ctx := context.WithValue(context.Background(), osc.ContextAWSv4, osc.AWSv4{
			AccessKey: d.Ak,
			SecretKey: d.Sk,
		})

		ctx = context.WithValue(ctx, osc.ContextServerIndex, 0)
		ctx = context.WithValue(ctx, osc.ContextServerVariables, map[string]string{"region": d.Region})

		d.oscApi = &OscApiData{
			client:  client,
			context: ctx,
		}
	}

	return d.oscApi, nil

}

// Create a host using the driver's config
func (d *OscDriver) Create() error {
	log.Debug("Creating a Vm")

	// Get the client
	oscApi, err := d.getClient()
	if err != nil {
		return err
	}

	// Create a keypair
	if err := createKeyPair(d); err != nil {
		cleanUp(d)
		return err
	}

	// Create a SG
	if d.securityGroupIds == nil {
		// Create default SG
		if err := createDefaultSecurityGroup(d); err != nil {
			cleanUp(d)
			return err
		}

		d.securityGroupIds = []string{d.SecurityGroupId}
	}

	// (TODO) Assign an Public IP
	if d.PublicCloud {
		if err := createPublicIp(d); err != nil {
			cleanUp(d)
			return err
		}
	}

	// Create an Instance
	deviceName := "/dev/sda1"
	rootDisk := osc.BlockDeviceMappingVmCreation{
		Bsu: &osc.BsuToCreate{
			VolumeType: &d.rootDiskType,
			VolumeSize: &d.rootDiskSize,
		},
		DeviceName: &deviceName,
	}

	if d.rootDiskType == "io1" {
		rootDisk.Bsu.SetIops(d.rootDiskIo1Iops)
	}

	createVmRequest := osc.CreateVmsRequest{
		ImageId:          d.sourceOmi,
		KeypairName:      &d.KeypairName,
		VmType:           &d.instanceType,
		SecurityGroupIds: &d.securityGroupIds,
		BlockDeviceMappings: &[]osc.BlockDeviceMappingVmCreation{
			rootDisk,
		},
	}

	if !d.PublicCloud {
		createVmRequest.SetSubnetId(d.subnetId)
	}

	var createVmResponse osc.CreateVmsResponse
	var httpRes *http.Response
	err = retry.Do(
		func() error {
			var response_error error
			createVmResponse, httpRes, response_error = oscApi.client.VmApi.CreateVms(oscApi.context).CreateVmsRequest(createVmRequest).Execute()
			return wrapError(response_error, httpRes)
		},
		defaultThrottlingRetryOption...,
	)

	if err != nil {
		cleanUp(d)
		return fmt.Errorf("Error while submitting the Vm creation request: %s", getErrorInfo(err, httpRes))
	}

	if !createVmResponse.HasVms() || len(createVmResponse.GetVms()) != 1 {
		cleanUp(d)
		return errors.New("Error while creating the Vm: the number of VM created is wrong")
	}

	// Store the VM Id
	d.VmId = createVmResponse.GetVms()[0].GetVmId()

	// Wait for the VM to be started
	log.Debug("Waiting for the Vm to be running...")
	if err := d.waitForState(d.VmId, "running"); err != nil {
		cleanUp(d)
		return errors.New("Error while waiting that the VM is running")
	}

	// Retrieve the Public IP
	readVmRequest := osc.ReadVmsRequest{
		Filters: &osc.FiltersVm{
			VmIds: &[]string{
				d.VmId,
			},
		},
	}

	var response osc.ReadVmsResponse
	err = retry.Do(
		func() error {
			var response_error error
			response, httpRes, response_error = oscApi.client.VmApi.ReadVms(oscApi.context).ReadVmsRequest(readVmRequest).Execute()
			return wrapError(response_error, httpRes)
		},
		defaultThrottlingRetryOption...,
	)
	if err != nil {
		cleanUp(d)
		return fmt.Errorf("Error while submitting the Vm read request: %s", getErrorInfo(err, httpRes))
	}

	if !response.HasVms() {
		cleanUp(d)
		return errors.New("Error while reading the VM: there is no VM")
	}

	if d.PublicCloud {
		// Link the Public Ip
		if err := linkPublicIp(d); err != nil {
			cleanUp(d)
			return err
		}
	} else {
		d.IPAddress = response.GetVms()[0].GetPrivateIp()
	}

	// Add the tag of the Vm name
	if err := addTag(d, d.VmId, "name", d.GetMachineName()); err != nil {
		cleanUp(d)
		return err
	}

	if d.tagK8sNodeName {
		// Add the tag of the Vm name
		if err := addTag(d, d.VmId, "OscK8sNodeName", d.GetMachineName()); err != nil {
			cleanUp(d)
			return err
		}
	}

	// Add extra tags to the Instances
	if err := addExtraTags(d, d.VmId, d.extraTagsAll); err != nil {
		cleanUp(d)
		return err
	}

	// Add extra tags only for the Instances
	if err := addExtraTags(d, d.VmId, d.extraTagsInstances); err != nil {
		cleanUp(d)
		return err
	}

	return nil
}

// DriverName returns the name of the driver
func (d *OscDriver) DriverName() string {
	return "outscale"
}

// GetCreateFlags returns the mcnflag.Flag slice representing the flags
// that can be set, their descriptions and defaults.
func (d *OscDriver) GetCreateFlags() []mcnflag.Flag {
	return []mcnflag.Flag{
		mcnflag.StringFlag{
			EnvVar: "OUTSCALE_ACCESS_KEY",
			Name:   flagAccessKey,
			Usage:  "Outscale Access Key",
			Value:  "",
		},
		mcnflag.StringFlag{
			EnvVar: "OUTSCALE_SECRET_KEY",
			Name:   flagSecretKey,
			Usage:  "Outscale Secret Key",
			Value:  "",
		},
		mcnflag.StringFlag{
			EnvVar: "OUTSCALE_REGION",
			Name:   flagRegion,
			Usage:  "Outscale Region (e.g. eu-west-2)",
			Value:  "",
		},
		mcnflag.StringFlag{
			EnvVar: "OUTSCALE_INSTANCE_TYPE",
			Name:   flagInstanceType,
			Usage:  "VM Instance type",
			Value:  defaultOscVmType,
		},
		mcnflag.StringFlag{
			EnvVar: "OUTSCALE_SOURCE_OMI",
			Name:   flagSourceOmi,
			Usage:  "OMI to use as bootstrap",
			Value:  defaultOscOMI,
		},
		mcnflag.StringSliceFlag{
			EnvVar: "",
			Name:   flagExtraTagsAll,
			Usage:  "Tags to set at all created resources",
			Value:  nil,
		},
		mcnflag.StringSliceFlag{
			EnvVar: "",
			Name:   flagExtraTagsInstances,
			Usage:  "Tags to set only to instances <key1=value1,key2=value2>",
			Value:  nil,
		},
		mcnflag.StringSliceFlag{
			EnvVar: "",
			Name:   flagSecurityGroupIds,
			Usage:  "Add machine into theses security groups",
			Value:  nil,
		},
		mcnflag.StringFlag{
			EnvVar: "",
			Name:   flagRootDiskType,
			Usage:  "Type of volume for the root disk ('standard', 'io1' or 'gp2')",
			Value:  defaultRootDiskType,
		},
		mcnflag.IntFlag{
			EnvVar: "",
			Name:   flagRootDiskSize,
			Usage:  "Size of the root disk in GB (> 0)",
			Value:  defaultRootDiskSize,
		},
		mcnflag.IntFlag{
			EnvVar: "",
			Name:   flagRootDiskIo1Iops,
			Usage:  "Iops for the io1 root disk type (ignore if it is not io1). Value between 1 and 13000.",
			Value:  defaultRootDiskIo1Iops,
		},
		mcnflag.StringFlag{
			EnvVar: "",
			Name:   flagSubnetId,
			Usage:  "Id of the Net use to create all resources when a private network is requested",
			Value:  "",
		},
		mcnflag.BoolFlag{
			EnvVar: "",
			Name:   flagK8sNodeNameTag,
			Usage:  "Automatically add kubernetes tag 'OscK8sNodeName' to the instance (Useful for the CCM)",
		},
	}
}

// GetIP returns an IP or hostname that this host is available at
// e.g. 1.2.3.4 or docker-host-d60b70a14d3a.cloudapp.net
func (d *OscDriver) GetIP() (string, error) {
	if d.IPAddress == "" {
		return "", errors.New("IP address is not set")
	}
	return d.IPAddress, nil
}

func (d *OscDriver) GetSSHHostname() (string, error) {
	return d.IPAddress, nil
}

// GetSSHPort returns port for use with ssh
func (d *OscDriver) GetSSHPort() (int, error) {
	return defaultSSHPort, nil
}

// GetSSHUsername returns username for use with ssh
func (d *OscDriver) GetSSHUsername() string {
	return defaultSSHUsername
}

// GetURL returns a Docker compatible host URL for connecting to this host
// e.g. tcp://1.2.3.4:2376
func (d *OscDriver) GetURL() (string, error) {
	return fmt.Sprintf("tcp://%s:%d", d.IPAddress, defaultDockerPort), nil
}

// GetState returns the state that the host is in (running, stopped, etc)
func (d *OscDriver) GetState() (state.State, error) {
	oscApi, err := d.getClient()
	if err != nil {
		return state.None, err
	}

	readVmRequest := osc.ReadVmsRequest{
		Filters: &osc.FiltersVm{
			VmIds: &[]string{
				d.VmId,
			},
		},
	}

	var readVmResponse osc.ReadVmsResponse
	var httpRes *http.Response
	err = retry.Do(
		func() error {
			var response_error error
			readVmResponse, httpRes, response_error = oscApi.client.VmApi.ReadVms(oscApi.context).ReadVmsRequest(readVmRequest).Execute()
			return wrapError(response_error, httpRes)
		},
		defaultThrottlingRetryOption...,
	)

	if err != nil {
		return state.None, fmt.Errorf("Error while submitting the Vm creation request: %s", getErrorInfo(err, httpRes))
	}

	if !readVmResponse.HasVms() {
		return state.None, errors.New("Error while reading the VM: there is no VM")
	}

	switch vmState := readVmResponse.GetVms()[0].GetState(); vmState {
	case "pending":
		return state.Starting, nil
	case "running":
		return state.Running, nil
	case "stopping", "shutting-down":
		return state.Stopping, nil
	case "stopped", "terminated", "quarantine":
		return state.Stopped, nil
	default:
		return state.None, nil
	}
}

// PreCreateCheck allows for pre-create operations to make sure a driver is ready for creation
func (d *OscDriver) PreCreateCheck() error {
	oscApi, err := d.getClient()
	if err != nil {
		return err
	}

	request := osc.ReadAccountsRequest{}

	var httpRes *http.Response
	err = retry.Do(
		func() error {
			var response_error error
			_, httpRes, response_error = oscApi.client.AccountApi.ReadAccounts(oscApi.context).ReadAccountsRequest(request).Execute()
			return wrapError(response_error, httpRes)
		},
		defaultThrottlingRetryOption...,
	)

	if err != nil {
		return fmt.Errorf("Error while submitting the ReadAccount request: %s", getErrorInfo(err, httpRes))
	}

	// Check the SG
	for _, sgId := range d.securityGroupIds {
		sgExist, sgError := isSecurityGroupExist(d, sgId)
		if sgError != nil {
			return sgError
		}

		if !sgExist {
			return fmt.Errorf("The Security Group '%v' does not exist.", sgId)
		}

		log.Debugf("The Security Group '%v' exists.", sgId)
	}

	// Check the SubnetId
	if !d.PublicCloud {
		netId, err := RetrieveNetFromSubnetId(d, d.subnetId)
		if err != nil {
			return err
		}

		if len(netId) == 0 {
			return fmt.Errorf("The Subnet Id '%v' does not exist.", d.subnetId)
		}

		d.netId = netId

		log.Debugf("The Subnet Id '%v' exists in NetId '%v'", d.subnetId, d.netId)
	}

	return nil

}

// Remove a host
func (d *OscDriver) Remove() error {
	oscApi, err := d.getClient()
	if err != nil {
		return err
	}

	if d.VmId != "" {
		request := osc.DeleteVmsRequest{
			VmIds: []string{
				d.VmId,
			},
		}

		var httpRes *http.Response
		err = retry.Do(
			func() error {
				var response_error error
				_, httpRes, response_error = oscApi.client.VmApi.DeleteVms(oscApi.context).DeleteVmsRequest(request).Execute()
				return wrapError(response_error, httpRes)
			},
			defaultThrottlingRetryOption...,
		)

		if err != nil {
			return fmt.Errorf("Error while submitting the DeleteVm request: %s", getErrorInfo(err, httpRes))
		}

		if err := d.waitForState(d.VmId, "terminated"); err != nil {
			return err
		}
	} else {
		log.Warn("Skipping deletion of the VM because none was stored.")
	}

	if err := deletePublicIp(d, d.PublicIpId); err != nil {
		return err
	}

	if err := deleteSecurityGroup(d, d.SecurityGroupId); err != nil {
		return err
	}

	if err := deleteKeyPair(d, d.KeypairName); err != nil {
		return err
	}

	return nil
}

// Restart a host. This may just call Stop(); Start() if the provider does not
// have any special restart behaviour.
func (d *OscDriver) Restart() error {
	oscApi, err := d.getClient()
	if err != nil {
		return err
	}

	request := osc.RebootVmsRequest{
		VmIds: []string{
			d.VmId,
		},
	}

	var httpRes *http.Response
	err = retry.Do(
		func() error {
			var response_error error
			_, httpRes, response_error = oscApi.client.VmApi.RebootVms(oscApi.context).RebootVmsRequest(request).Execute()
			return wrapError(response_error, httpRes)
		},
		defaultThrottlingRetryOption...,
	)

	if err != nil {
		return fmt.Errorf("Error while submitting the RebootVm request: %s", getErrorInfo(err, httpRes))
	}

	if err := d.waitForState(d.VmId, "running"); err != nil {
		return err
	}

	return nil
}

// SetConfigFromFlags configures the driver with the object that was returned
// by RegisterCreateFlags
func (d *OscDriver) SetConfigFromFlags(flags drivers.DriverOptions) error {
	if d.Ak = flags.String(flagAccessKey); d.Ak == "" {
		if d.Ak = os.Getenv("OSC_ACCESS_KEY"); d.Ak == "" {
			return errors.New("Outscale Access Key is required")
		}
	}

	if d.Sk = flags.String(flagSecretKey); d.Sk == "" {
		if d.Sk = os.Getenv("OSC_SECRET_KEY"); d.Ak == "" {
			return errors.New("Outscale Secret key is required")
		}
	}

	if d.Region = flags.String(flagRegion); d.Region == "" {
		if d.Region = os.Getenv("OSC_REGION"); d.Region == "" {
			d.Region = defaultOscRegion
		}
	}

	d.instanceType = flags.String(flagInstanceType)
	d.sourceOmi = flags.String(flagSourceOmi)

	// Root disk
	d.rootDiskType = flags.String(flagRootDiskType)
	if !validateDiskType(d.rootDiskType) {
		return fmt.Errorf("the disk type is not accepted (got: %s, expected: 'standard'|'io1'|'gp2')", d.rootDiskType)
	}
	if d.rootDiskSize = int32(flags.Int(flagRootDiskSize)); d.rootDiskSize <= 0 {
		return fmt.Errorf("the disk size (%v) is not accepted, it must be > 0", d.rootDiskSize)
	}

	if d.rootDiskIo1Iops = int32(flags.Int(flagRootDiskIo1Iops)); d.rootDiskIo1Iops <= 0 {
		return fmt.Errorf("the disk iops (%v) is not accepted, it must between 1 and 13000", d.rootDiskIo1Iops)
	}

	// Tags
	if d.extraTagsAll = flags.StringSlice(flagExtraTagsAll); !validateExtraTagsFormat(d.extraTagsAll) {
		return fmt.Errorf("--%v have not the expected syntax", flagExtraTagsAll)
	}

	if d.extraTagsInstances = flags.StringSlice(flagExtraTagsInstances); !validateExtraTagsFormat(d.extraTagsInstances) {
		return fmt.Errorf("--%v have not the expected syntax", flagExtraTagsInstances)
	}

	d.tagK8sNodeName = flags.Bool(flagK8sNodeNameTag)

	// Security Groups
	d.securityGroupIds = flags.StringSlice(flagSecurityGroupIds)

	// Private or Public Cloud
	d.subnetId = flags.String(flagSubnetId)
	d.PublicCloud = len(d.subnetId) == 0

	// SSH
	d.SSHKeyPath = d.GetSSHKeyPath()
	d.SSHUser = d.GetSSHUsername()
	d.SSHPort, _ = d.GetSSHPort()

	return nil
}

// Start a host
func (d *OscDriver) Start() error {
	oscApi, err := d.getClient()
	if err != nil {
		return err
	}

	request := osc.StartVmsRequest{
		VmIds: []string{
			d.VmId,
		},
	}

	var httpRes *http.Response
	err = retry.Do(
		func() error {
			var response_error error
			_, httpRes, response_error = oscApi.client.VmApi.StartVms(oscApi.context).StartVmsRequest(request).Execute()
			return wrapError(response_error, httpRes)
		},
		defaultThrottlingRetryOption...,
	)

	if err != nil {
		return fmt.Errorf("Error while submitting the StartVm request: %s", getErrorInfo(err, httpRes))
	}

	if err := d.waitForState(d.VmId, "running"); err != nil {
		return err
	}

	return nil

}

func (d *OscDriver) innerStop(force bool) error {
	oscApi, err := d.getClient()
	if err != nil {
		return err
	}

	request := osc.StopVmsRequest{
		VmIds: []string{
			d.VmId,
		},
	}
	request.SetForceStop(force)

	var httpRes *http.Response
	err = retry.Do(
		func() error {
			var response_error error
			_, httpRes, response_error = oscApi.client.VmApi.StopVms(oscApi.context).StopVmsRequest(request).Execute()
			return wrapError(response_error, httpRes)
		},
		defaultThrottlingRetryOption...,
	)

	if err != nil {
		return fmt.Errorf("Error while submitting the StopVm request: %s", getErrorInfo(err, httpRes))
	}

	if err := d.waitForState(d.VmId, "stopped"); err != nil {
		return err
	}

	return nil
}

// Stop a host gracefully
func (d *OscDriver) Stop() error {
	return d.innerStop(false)
}

// Kill stops a host forcefully
func (d *OscDriver) Kill() error {
	return d.innerStop(true)
}

func validateDiskType(diskType string) bool {
	switch diskType {
	case "io1", "gp2", "standard":
		return true
	default:
		return false
	}
}
